doc/kernel-limitations.md

```diff
@@ -0,0 +1,169 @@
+# Notable Kernel limitations in the current Mystikos implementation
+
+The first hardware TEE that Mystikos based on is Intel SGX. Therefore, Mystikos
+carries most limitations imposed by Intel SGX. However, we expect those
+limitations can be relaxed as better hardware TEE platforms emerge, or when
+Mystikos overcomes a hardware limitation with software emulation. In fact,
+[SGX2](https://caslab.csl.yale.edu/workshops/hasp2016/HASP16-16_slides.pdf)
+has already relaxed a few limitations imposed by SGX1, as shown below.
+
+By and large, Mystikos implements a Linux-compatible kernel so it could
+run Linux applications with little or no modifications. There are still
+incompatibilities/limitations worth noting as summarized below:
+
+| Limitation                    | Description         | Potential impacts on applications |
+| ------------------------------|---------------------|---------|
+| [Multi-process limitation](#limitations-arisen-from-sgxs-single-process-model) | No forking of child processes. However `posix_spawn` is supported. | Communication and address space isolation between processes |
+| [Thread limitations](#limitations-arisen-from-sgx1-limitations) | Thread creation is slow and limited to a user setting. Thread scheduling is subject to attacks by host. | Required Configuration / Thread starvation / D.o.S. |
+| [Memory limitations](#limitations-arisen-from-sgx1-limitations) | User stack/heap memory is limited to a user setting. Page permission enforcement is under host control. | Required Configuration / OOM / Page table manipulation |
+| [Clock limitation](#limitations-arisen-from-lack-of-access-to-time-source) | Clock ticks and resolution are controlled by host | Clock resolution / Untrusted time |
+| [Exception handling limitation](#hardware-exception-handling-limitations) | Direct #PF/#GP exception handling inside the Mystikos environment is not supported | Missed or arbitrary exceptions / Exception handling with long delays |
```

Comment:
```suggestion
| [Exception handling limitation](#limitations-arisen-from-sgx1-limitations) | Direct #PF/#GP exception handling inside the Mystikos environment is not supported |  Exception handling with long delays |
```
#FP/#GP exception will not be missed. If exception is triggered either based on OS/VMM page table or SGX EPCM, the SGX enclave will go through SGX AEX and the Enclave execution will not make forward progress even if the OS does not invoke the trusted exception handler inside the Enclave. If the OS invokes the trusted exception handler, the handler will abort Enclave execution, when it detects that the trusted exception info is not present inside the Enclave. "Arbitrary exception" terminology does not paint the full picture. The system SW can cause interrupt or #PF exception, which typically is a security issue for TEE. See discussion in another comment.